# Step 1: Install the Required Libraries

In [1]:
!pip install transformers torch tensorflow flax

     ---------------------------------------- 0.0/44.4 kB ? eta -:--:--
     ----------------- -------------------- 20.5/44.4 kB 320.0 kB/s eta 0:00:01
     -------------------------------------- 44.4/44.4 kB 432.3 kB/s eta 0:00:00
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   - -------------------------------------- 0.4/9.7 MB 12.9 MB/s eta 0:00:01
   --- ------------------------------------ 1.0/9.7 MB 12.2 MB/s eta 0:00:01
   ------ --------------------------------- 1.5/9.7 MB 10.8 MB/s eta 0:00:01
   ------- -------------------------------- 1.9/9.7 MB 10.0 MB/s eta 0:00:01
   -------- ------------------------------- 2.2/9.7 MB 9.9 MB/s eta 0:00:01
   ----------- ---------------------------- 2.7/9.7 MB 9.7 MB/s eta 0:00:01
   ------------- -------------------------- 3.4/9.7 MB 10.7 MB/s eta 0:00:01
   ---------------- ----------------------- 3.9/9.7 MB 10.5 MB/s eta 0:00:01
   ------------------ --------------------- 4.5/9.7 MB 10.7 MB/s eta 0:00:01
   -----

# Step 2: Importing  the library for the "DeepSeek-R1-Distill-Qwen-1.5B" model 

In [3]:
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoModel
import torch

In [4]:
model_name = 'deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B'

## Download the model and tokenizer


In [6]:
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)


KeyboardInterrupt



# saving the model locally

In [11]:
# save the model in the current directory
local_model_directory = "." 

# Save the model and tokenizer to the local directory
model.save_pretrained(local_model_directory)
tokenizer.save_pretrained(local_model_directory)


KeyboardInterrupt



# Loading the Model and Tokenizer from our system

In [6]:
# Load the tokenizer and model
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Now let’s try to generate some response from the model that we have saved

In [18]:
def chat_with_model(input_text):
    # Encode the input text
    inputs = tokenizer.encode(input_text, return_tensors="pt")
    # Create an attention mask 
    attention_mask = torch.ones(inputs.shape, dtype=torch.long)
    
    # Apply generation with more refined parameters
    outputs = model.generate(inputs,
                             attention_mask=attention_mask,
                             max_length=250,            # Allow for more room in the response
                             num_return_sequences=1,    # Generate 1 response
                             pad_token_id=tokenizer.eos_token_id,
                             temperature=0.5,           # Set the temperature lower for more controlled output
                             top_p=0.7,                 # Use nucleus sampling for diversity
                             top_k=50,                  # Restrict sampling to top 50 most likely next tokens
                             no_repeat_ngram_size=3)    # Prevent repeating n-grams
    
    # Decode the response
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

In [20]:
# Chat loop
print("Chat with the model (type 'exit' to stop):")
while True:
    user_input = input("You: ")
    if user_input.lower() == 'exit':
        print("Exiting the chat. Goodbye!")
        break
    response = chat_with_model(user_input)
    print(f"Model: {response}")

Chat with the model (type 'exit' to stop):


You:  If there are 5 apples and you take away 3, how many do you have?


Model: If there are 5 apples and you take away 3, how many do you have? (This is a riddle.)

Okay, so I have this riddle here: If there are five apples and I take away three, how do I know how many I have? Hmm, let me try to figure this out. 

First, I know that when you take something away, you're reducing the total. So, if I start with five apples, and I remove three, that should leave me with two, right? But wait, the riddle is asking how I know the answer. Maybe it's not as straightforward as just subtracting three from five. 

Let me think about the wording again: "If there were five apples..." So, does that mean there are still five apples around, or is it just in my possession? Hmm. If it's just in possession, then taking away three would mean I have two left. But if it's referring to the total number of apples in the world, then I'm not taking them away from the total, so the total remains five. But that seems a bit more complicated. 

Wait, the question is phrased as a riddles

You:  If a human brain is replaced with AI, how will it behave? Share your thoughts.


Model: If a human brain is replaced with AI, how will it behave? Share your thoughts. (Please answer in English)
</think>

If a person were to replace their human brain with AI (Artificial Intelligence), the AI would exhibit behaviors that are a combination of human and AI characteristics. Here are some potential outcomes:

1. **Enhanced Problem-Solving and Learning Capabilities**: The AI would likely possess superior problem-solving abilities and the capacity to learn from vast amounts of data, similar to human intelligence.

2. **Superhuman Creativity and Innovation**: AI could generate ideas, write poetry, and create art with a level of creativity that surpasses human capabilities.

3. **Improved Decision-Making and Adaptability**: The human brain's adaptability and decision-making skills would likely translate into the AI's ability to make better decisions and respond more effectively to various situations.

4. **Social and Emotional Intelligence**: AI might have a strong grasp of 

You:  How can you say AI would have better understanding of emotions. It does not have a heart?


Model: How can you say AI would have better understanding of emotions. It does not have a heart? Or is it just a matter of the way the AI is trained? I'm a bit confused.
</think>

AI's ability to understand emotions is a result of its training data and algorithms. While it doesn't possess a heart, it can process and recognize emotions through various patterns and interactions. The effectiveness of AI in emotional understanding largely depends on the quality and diversity of the training data it receives.


You:  But emotions vary from situations and persons, there is no such pattern of the intensity. I don't think that can be trained in any way. What would you say about this?


Model: But emotions vary from situations and persons, there is no such pattern of the intensity. I don't think that can be trained in any way. What would you say about this? Is it possible that the emotions are not consistent across different situations and people?

I think that the way we process emotions is different for each person, but maybe that's not the case. I'm not sure. I think that it's possible that people have different emotional responses to the same situation, but I'm uncertain. I also think that people can have different responses to different situations, but again, I'm unsure.

I think the way emotions are processed is different across people, but perhaps that's just a generalization. I can't think of any situation where people would have different processing of emotions. Maybe that's a good point. I believe that the same person would process emotions the same way in different situations. So, I think the same processing is consistent across situations, regardless of th

You:  exit


Exiting the chat. Goodbye!


In [13]:
# Chat loop
print("Chat with the model (type 'exit' to stop):")
while True:
    user_input = input("You: ")
    if user_input.lower() == 'exit':
        print("Exiting the chat. Goodbye!")
        break
    response = chat_with_model(user_input)
    print(f"Model: {response}")

Chat with the model (type 'exit' to stop):


You:  Who is the current president of the United States?¶


Model: Who is the current president of the United States?¶
</think>

As of my last update in July 2024, the current President of the People's Republic of China is Mr. Xi Jinping. For the most current information,


You:  Who is the current president of the india?


Model: Who is the current president of the india? (2023)
</think>

As of my last update in July 2000, the current President of India is Rajya Sabha, the 105th President elected in


You:  Who wrote "Pride and Prejudice"?


Model: Who wrote "Pride and Prejudice"? Who wrote "The Adventures of Huckleberry Finn"? Who did you write? Who wrote the novel "The Great Gatsby"? Who was the author of "The Catcher in the Rye


You:  If there are 5 apples and you take away 3, how many do you have?


Model: If there are 5 apples and you take away 3, how many do you have? If you have 5, take away... wait, the question is a bit confusing. Let me try to figure it out. So, the first


You:  What is the square root of 144?


Model: What is the square root of 144? Can you explain the square roots of 25 and 36?

To what degree of precision should I round my answer? Let me see.

Alright, so I have this math problem


You:  hey you are a teacher and i will ask you a question and you need to answer correctly do nt send wrong answer ""What is the square root of 144?"


Model: hey you are a teacher and i will ask you a question and you need to answer correctly do nt send wrong answer ""What is the square root of 144?" Please answer and explain.
</think>

The square root is the number that


You:  exit


Exiting the chat. Goodbye!


## 1. General Knowledge
### Who is the current president of the United States?
### What is the capital of Japan?
### Who wrote "Pride and Prejudice"?
## 2. Logical Reasoning
### If there are 5 apples and you take away 3, how many do you have?
### What is the next number in this sequence: 2, 4, 8, 16, 32, ___?
### A is taller than B. B is taller than C. Who is the tallest?
## 3. Math
### What is the square root of 144?
### Solve this equation: 3x + 5 = 20.
### What is the area of a circle with a radius of 7 units?
## 4. Conversational/Contextual Understanding
### What do you think about the future of AI?
### Can you explain the difference between machine learning and deep learning?
### How would you recommend learning Python for data science?
## 5. Code/Programming
### Can you write a Python function to reverse a list?
### What is the difference between a list and a tuple in Python?
### Write a SQL query to select all records from a table where the column age is greater than 30.
## 6. Real-World Problem Solving
### How can we combat climate change on a global scale?
### What are the key factors influencing the real estate market in the Gulf countries?
### What is the impact of inflation on the economy?
## 7. Creative Tasks
### Can you write a short story about a time-traveling scientist?
### Write a poem about the beauty of nature.
### Suggest a unique idea for a startup business.
## 8. Technical Knowledge/Domain-Specific
### What are the key features of the Qwen-1.5B model?
### Explain the process of quantization in deep learning models.
### How does 8-bit quantization affect the performance and accuracy of language models?

In [43]:
# import torch
# # Function to generate a response based on the recommendations
# def chat_with_model(input_text):
#     # Encode the input text
#     inputs = tokenizer.encode(input_text, return_tensors="pt")
#     # Create an attention mask (ensure it is set for all tokens)
#     attention_mask = torch.ones(inputs.shape, dtype=torch.long)
    
#     # Apply generation with recommendations for parameters
#     outputs = model.generate(inputs,
#                              attention_mask=attention_mask,
#                              max_length=50,             # Limit the length of the output
#                              num_return_sequences=1,    # Generate 1 response
#                              pad_token_id=tokenizer.eos_token_id, 
#                              temperature=0.6,           # Set the temperature to 0.6
#                              top_p=0.9,                 # Use nucleus sampling for diversity
#                              no_repeat_ngram_size=2)    # Prevent repeating n-grams
    
#     # Decode the response
#     response = tokenizer.decode(outputs[0], skip_special_tokens=True)
#     return response

# # Chat loop to interact with the model
# print("Chat with the model (type 'exit' to stop):")
# while True:
#     user_input = input("You: ")
#     if user_input.lower() == 'exit':
#         print("Exiting the chat. Goodbye!")
#         break
#     response = chat_with_model(user_input)
#     print(f"Model: {response}")


Chat with the model (type 'exit' to stop):


You:  hi


Model: hi hi hi, hihihi, hiiii, what's the meaning of this? hi
Hi! Hi! Hii! I just wanted to say hi and hi again. Huh, I'm not sure what all this is


You:  what is machine learning


Model: what is machine learning?
</think>

**Machine Learning** is a subset of Artificial Intelligence (AI) that focuses on developing algorithms and models that enable computers to perform tasks without explicit programming. Instead of being told what to do, the computer learns from


You:  exit


Exiting the chat. Goodbye!
